handel/src/aggregation.rs

```diff
@@ -430,98 +526,132 @@ impl<
+    /// Used to deconstruct the OngoingAggregation and should be `unreachable!()` everywhere but in that location.
+    Transitioning,
 }
 
-impl<
-        TId: Debug + Clone + Unpin + Send + 'static,
-        TProtocol: Protocol<TId>,
-        TNetwork: Network<Contribution = TProtocol::Contribution> + Send + 'static,
-    > Aggregation<TId, TProtocol, TNetwork>
+impl<TId, P, N> Aggregation<TId, P, N>
+where
+    TId: Debug + Clone + Unpin + Send + 'static,
+    P: Protocol<TId>,
+    N: Network<Contribution = P::Contribution>,
 {
     pub fn new(
-        protocol: TProtocol,
+        protocol: P,
         config: Config,
-        own_contribution: TProtocol::Contribution,
-        input_stream: LevelUpdateStream<TProtocol, TId>,
-        network: TNetwork,
+        own_contribution: P::Contribution,
+        input_stream: LevelUpdateStream<P, TId>,
+        network: N,
     ) -> Self {
         // Create the Sender, buffering a single message per recipient.
         let sender = LevelUpdateSender::new(protocol.partitioner().size(), network);
 
-        let next_aggregation =
-            NextAggregation::new(protocol, config, own_contribution, input_stream, sender)
-                .next()
-                .boxed();
-
-        Self {
-            next_aggregation: Some(next_aggregation),
-        }
+        // Aggregations start out as Ongoing
+        Self::Ongoing(OngoingAggregation::new(
+            protocol,
+            config,
+            own_contribution,
+            input_stream,
+            sender,
+        ))
     }
 }
 
-impl<
-        TId: Debug + Clone + Unpin + Send + 'static,
-        P: Protocol<TId> + Debug,
-        N: Network<Contribution = P::Contribution>,
-    > Stream for Aggregation<TId, P, N>
+impl<TId, P, N> Stream for Aggregation<TId, P, N>
+where
+    TId: Debug + Clone + Unpin + Send + 'static,
+    P: Protocol<TId>,
+    N: Network<Contribution = P::Contribution>,
 {
     type Item = P::Contribution;
 
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        // check if there is a next_aggregation
-        let next_aggregation = match self.next_aggregation.as_mut() {
-            // If there is Some(next_aggregation) proceed with it
-            Some(next_aggregation) => next_aggregation,
-            // If there is None that means all signatories have signed the payload so there is nothing more to aggregate after the last returned value.
-            // Thus return Poll::Ready(None) as the previous value can no longer be improved.
-            None => return Poll::Ready(None),
+        // Poll either the ongoing or the finished aggregation
+        let result = match &mut *self {
+            Self::Transitioning => {
+                unreachable!("Aggregation should never be transitioning when polled.")
+            }
+            Self::Finished(ref mut finished_aggregation) => {
```

Comment:
```suggestion
            Self::Finished(finished_aggregation) => {
```
Here and below, I thought the `ref mut` is unnecessary, but maybe I'm wrong.

Reply:
Works 👍!